test/gherkin.go

```diff
@@ -12,195 +14,161 @@ const (
+	slices.ForEach(thens, func(thens ThenStatements) { checkFirstWordOfLabels(thens, assertNotGIVEN) })
+
+	out := ThenStatements{}
+	for _, then := range thens {
+		for _, statement := range then {
+			statement := statement
+			out = append(out, thenStatement{
+				label: mergeLabels(w.label, statement.label),
+				test:  func(t *testing.T) { w.test(); statement.test(t) },
+			})
+		}
 	}
+
+	return out
 }
 
-// Then adds an outcome check to the test path
-func (t then) Then(description string, execution func(t *testing.T)) Runner {
-	return runner{
-		label: append(t.label, _then, description),
-		test:  append(t.test, execution),
+// When adds a trigger to the test path
+func (w WhenStatement) When(description string, setup func()) WhenStatement {
+	return WhenStatement{
+		label: append(w.label, _and, _when, description),
+		test:  func() { w.test(); setup() },
 	}
 }
 
-// ForEach generates a test for each test case
-func (tc testCases[T]) ForEach(createRunner func(testCase T) Runner) Runner {
-
-	runners := make([]Runner, len(tc.items))
-
-	for i, v := range tc.items {
-		runners[i] = createRunner(v)
+func (w WhenStatement) When2(w2 WhenStatement) WhenStatement {
+	return WhenStatement{
+		label: mergeLabels(w.label, w2.label),
+		test:  func() { w.test(); w2.test() },
 	}
-
-	return branch([]string{}, []func(t *testing.T){}, runners)
 }
 
-// Run executes all defined test paths. Optionally, each path is repeated a given number of times
-func (r multiRunner) Run(t *testing.T, repeats ...int) bool {
-	result := true
-	for _, runner := range r.runners {
-		// cannot inline this because if result is false the second part of && is not going to be evaluated
-		newResult := runner.Run(t, repeats...)
-		result = result && newResult
-	}
-	return result
+// Then adds an outcome check to the test path
+func (w WhenStatement) Then(description string, execution func(t *testing.T)) ThenStatements {
+	return []thenStatement{{
+		label: append(w.label, _then, description),
+		test:  func(t *testing.T) { w.test(); execution(t) },
+	}}
+}
+
+// Then2 allows the use of a previously defined "then" statement
+func (w WhenStatement) Then2(then ThenStatements) ThenStatements {
+	return slices.Map(then, func(then thenStatement) thenStatement {
+		return thenStatement{
+			label: mergeLabels(w.label, then.label),
+			test:  func(t *testing.T) { w.test(); then.test(t) },
+		}
+	})
 }
 
 // Run executes all defined test paths. Optionally, each path is repeated a given number of times
-func (r runner) Run(t *testing.T, repeats ...int) bool {
+func (then thenStatement) Run(t *testing.T, repeats ...int) bool {
 	repeat := 1
 	if len(repeats) == 1 && repeats[0] > 1 {
 		repeat = repeats[0]
 	}
-	return t.Run(strings.Join(r.label, " "), Func(func(t *testing.T) {
-		for _, f := range r.test {
-			f(t)
-		}
-	}).Repeat(repeat))
+	return t.Run(strings.Join(then.label, " "), Func(then.test).Repeat(repeat))
```

Comment:
Another unrelated thing, I think it's better to have `.Repeat` stop immediately once any of the tests fail. The console typically looks messy and hard to understand what's happening when, say 100, instances of the same test all reports failure in one shot.